source/connection_monitor.c

```diff
@@ -0,0 +1,217 @@
+        if (aws_array_list_get_at(stats_list, &stats_base, i)) {
+            continue;
+        }
+
+        switch (stats_base->category) {
+            case AWSCRT_STAT_CAT_SOCKET: {
+                struct aws_crt_statistics_socket *socket_stats = (struct aws_crt_statistics_socket *)stats_base;
+                bytes_read = socket_stats->bytes_read;
+                bytes_written = socket_stats->bytes_written;
+                break;
+            }
+
+            case AWSCRT_STAT_CAT_HTTP1_CHANNEL: {
+                struct aws_crt_statistics_http1_channel *http1_stats =
+                    (struct aws_crt_statistics_http1_channel *)stats_base;
+                pending_read_interval_ms = http1_stats->pending_incoming_stream_ms;
+                pending_write_interval_ms = http1_stats->pending_outgoing_stream_ms;
+                current_outgoing_stream_id = http1_stats->current_outgoing_stream_id;
+                current_incoming_stream_id = http1_stats->current_incoming_stream_id;
+
+                break;
+            }
+
+            default:
+                break;
+        }
+    }
+
+    struct aws_channel *channel = context;
+
+    uint64_t bytes_per_second = 0;
+    uint64_t max_pending_io_interval_ms = 0;
+
+    if (pending_write_interval_ms > 0) {
+        double fractional_bytes_written_per_second =
+            (double)bytes_written * (double)AWS_TIMESTAMP_MILLIS / (double)pending_write_interval_ms;
+        if (fractional_bytes_written_per_second >= (double)UINT64_MAX) {
+            bytes_per_second = UINT64_MAX;
+        } else {
+            bytes_per_second = (uint64_t)fractional_bytes_written_per_second;
+        }
+        max_pending_io_interval_ms = pending_write_interval_ms;
+    }
+
+    if (pending_read_interval_ms > 0) {
+        double fractional_bytes_read_per_second =
+            (double)bytes_read * (double)AWS_TIMESTAMP_MILLIS / (double)pending_read_interval_ms;
+        if (fractional_bytes_read_per_second >= (double)UINT64_MAX) {
+            bytes_per_second = UINT64_MAX;
+        } else {
+            bytes_per_second = aws_add_u64_saturating(bytes_per_second, (uint64_t)fractional_bytes_read_per_second);
+        }
+        if (pending_read_interval_ms > max_pending_io_interval_ms) {
+            max_pending_io_interval_ms = pending_read_interval_ms;
+        }
+    }
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_IO_CHANNEL,
+        "id=%p: channel throughput - %" PRIu64 " bytes per second",
+        (void *)channel,
+        bytes_per_second);
+
+    /* Check throughput only if at least one stream exists and was observed in that role pregviously */
+    bool check_throughput =
+        (current_incoming_stream_id != 0 && current_incoming_stream_id == impl->last_incoming_stream_id) ||
+        (current_outgoing_stream_id != 0 && current_outgoing_stream_id == impl->last_outgoing_stream_id);
+
+    impl->last_outgoing_stream_id = current_outgoing_stream_id;
```

Comment:
how does this handle parallel streams? wouldn't we need a monitor per stream in that case?

Reply:
When we update this to support H2, this particular logic would only apply to h1 streams.  H2 streams would have their own stats structure which we can configure to contain the data we need to make a similar condition (if that makes sense) or for h2 streams we could always just drop this requirement.  

The fact that the streams have a simple monotonic id mechanism could mean that just keeping track of the current min and max within a connection is sufficient to apply the same conditional filter (ie if the current min id is less than the previous max id then you know at least one stream exists from the previous sample timepoint).